File: plugins/org.eclipse.tm.tcf.core/src/org/eclipse/tm/internal/tcf/services/remote/FileSystemProxy.java
/*******************************************************************************
 * Copyright (c) 2007, 2008 Wind River Systems, Inc. and others.
 * All rights reserved. This program and the accompanying materials 
 * are made available under the terms of the Eclipse Public License v1.0 
 * which accompanies this distribution, and is available at 
 * http://www.eclipse.org/legal/epl-v10.html 
 *  
 * Contributors:
 *     Wind River Systems - initial API and implementation
 *******************************************************************************/
package org.eclipse.tm.internal.tcf.services.remote;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import org.eclipse.tm.tcf.core.Command;
import org.eclipse.tm.tcf.protocol.IChannel;
import org.eclipse.tm.tcf.protocol.IErrorReport;
import org.eclipse.tm.tcf.protocol.IToken;
import org.eclipse.tm.tcf.protocol.JSON;
import org.eclipse.tm.tcf.services.IFileSystem;


public class FileSystemProxy implements IFileSystem {
    
    private final class FileHandle implements IFileHandle {
        final String id;
        
        FileHandle(String id) {
            this.id = id;
        }

        public IFileSystem getService() {
            return FileSystemProxy.this;
        }
        
        public String toString() {
            return "[File Handle '" + id + "']";
        }
    }
    
    private static final class Status extends FileSystemException implements IErrorReport {
        
        private static final long serialVersionUID = -1636567076145085980L;
        
        private final int status;
        private final Map<String,Object> attrs;
        
        Status(int status, String message, Map<String,Object> attrs) {
            super(message);
            this.status = status;
            this.attrs = attrs;
        }
        
        Status(Exception x) {
            super(x);
            this.status = IErrorReport.TCF_ERROR_OTHER;
            this.attrs = new HashMap<String,Object>();
        }
        
        public int getStatus() {
            return status;
        }

        public int getErrorCode() {
            Number n = (Number)attrs.get(ERROR_CODE);
            if (n == null) return 0;
            return n.intValue();
        }

        public int getAltCode() {
            Number n = (Number)attrs.get(ERROR_ALT_CODE);
            if (n == null) return 0;
            return n.intValue();
        }

        public String getAltOrg() {
            return (String)attrs.get(ERROR_ALT_ORG);
        }

        public Map<String, Object> getAttributes() {
            return attrs;
        }
    }
    
    private abstract class FileSystemCommand extends Command {
        
        FileSystemCommand(String command, Object[] args) {
            super(channel, FileSystemProxy.this, command, args);
        }
        
        @SuppressWarnings("unchecked")
        public Status toFSError(Object data) {
            if (data == null) return null;
            Map<String,Object> map = (Map<String,Object>)data;
            Number error_code = (Number)map.get(IErrorReport.ERROR_CODE);
            String cmd = getCommandString();
            if (cmd.length() > 72) cmd = cmd.substring(0, 72) + "...";
            Status s = new Status(error_code.intValue(),
                    "TCF command exception:" +
                    "\nCommand: " + cmd +
                    "\nException: " + toErrorString(data) +
                    "\nError code: " + error_code, map);
            Object caused_by = map.get(IErrorReport.ERROR_CAUSED_BY);
            if (caused_by != null) s.initCause(toError(caused_by, false));
            return s;
        }
    }
    
    private final IChannel channel;
    
    public FileSystemProxy(IChannel channel) {
        this.channel = channel;
    }

    public IToken close(IFileHandle handle, final DoneClose done) {
        assert handle.getService() == this;
        String id = ((FileHandle)handle).id;
        return new FileSystemCommand("close", new Object[]{ id }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 1;
                    s = toFSError(args[0]);
                }
                done.doneClose(token, s);
            }
        }.token;
    }

    public IToken setstat(String path, FileAttrs attrs, final DoneSetStat done) {
        Object dt = toObject(attrs);
        return new FileSystemCommand("setstat", new Object[]{ path, dt }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 1;
                    s = toFSError(args[0]);
                }
                done.doneSetStat(token, s);
            }
        }.token;
    }

    public IToken fsetstat(IFileHandle handle, FileAttrs attrs, final DoneSetStat done) {
        assert handle.getService() == this;
        String id = ((FileHandle)handle).id;
        Object dt = toObject(attrs);
        return new FileSystemCommand("fsetstat", new Object[]{ id, dt }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 1;
                    s = toFSError(args[0]);
                }
                done.doneSetStat(token, s);
            }
        }.token;
    }

    public IToken stat(String path, final DoneStat done) {
        return new FileSystemCommand("stat", new Object[]{ path }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                FileAttrs a = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 2;
                    s = toFSError(args[0]);
                    if (s == null) a = toFileAttrs(args[1]);
                }
                done.doneStat(token, s, a);
            }
        }.token;
    }

    public IToken fstat(IFileHandle handle, final DoneStat done) {
        assert handle.getService() == this;
        String id = ((FileHandle)handle).id;
        return new FileSystemCommand("fstat", new Object[]{ id }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                FileAttrs a = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 2;
                    s = toFSError(args[0]);
                    if (s == null) a = toFileAttrs(args[1]);
                }
                done.doneStat(token, s, a);
            }
        }.token;
    }

    public IToken lstat(String path, final DoneStat done) {
        return new FileSystemCommand("lstat", new Object[]{ path }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                FileAttrs a = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 2;
                    s = toFSError(args[0]);
                    if (s == null) a = toFileAttrs(args[1]);
                }
                done.doneStat(token, s, a);
            }
        }.token;
    }

    public IToken mkdir(String path, FileAttrs attrs, final DoneMkDir done) {
        Object dt = toObject(attrs);
        return new FileSystemCommand("mkdir", new Object[]{ path, dt }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 1;
                    s = toFSError(args[0]);
                }
                done.doneMkDir(token, s);
            }
        }.token;
    }

    public IToken open(String file_name, int flags, FileAttrs attrs, final DoneOpen done) {
        Object dt = toObject(attrs);
        return new FileSystemCommand("open", new Object[]{ file_name, new Integer(flags), dt }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                FileHandle h = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 2;
                    s = toFSError(args[0]);
                    if (s == null) h = toFileHandle(args[1]);
                }
                done.doneOpen(token, s, h);
            }
        }.token;
    }

    public IToken opendir(String path, final DoneOpen done) {
        return new FileSystemCommand("opendir", new Object[]{ path }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                FileHandle h = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 2;
                    s = toFSError(args[0]);
                    if (s == null) h = toFileHandle(args[1]);
                }
                done.doneOpen(token, s, h);
            }
        }.token;
    }

    public IToken read(IFileHandle handle, long offset, int len, final DoneRead done) {
        assert handle.getService() == this;
        String id = ((FileHandle)handle).id;
        return new FileSystemCommand("read", new Object[]{
                id, Long.valueOf(offset), Integer.valueOf(len) }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                byte[] b = null;
                boolean eof = false;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 3;
                    s = toFSError(args[1]);
                    if (s == null) {
                        b = JSON.toByteArray(args[0]);
                        eof = ((Boolean)args[2]).booleanValue();
                    }
                }
                done.doneRead(token, s, b, eof);
            }
        }.token;
    }

    public IToken readdir(IFileHandle handle, final DoneReadDir done) {
        assert handle.getService() == this;
        String id = ((FileHandle)handle).id;
        return new FileSystemCommand("readdir", new Object[]{ id }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                DirEntry[] b = null;
                boolean eof = false;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 3;
                    s = toFSError(args[1]);
                    if (s == null) {
                        b = toDirEntryArray(args[0]);
                        eof = ((Boolean)args[2]).booleanValue();
                    }
                }
                done.doneReadDir(token, s, b, eof);
            }
        }.token;
    }

    public IToken roots(final DoneRoots done) {
        return new FileSystemCommand("roots", null) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                DirEntry[] b = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 2;
                    s = toFSError(args[1]);
                    if (s == null) b = toDirEntryArray(args[0]);
                }
                done.doneRoots(token, s, b);
            }
        }.token;
    }

    public IToken readlink(String path, final DoneReadLink done) {
        return new FileSystemCommand("readlink", new Object[]{ path }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                String p = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 2;
                    s = toFSError(args[0]);
                    if (s == null) p = (String)args[1];
                }
                done.doneReadLink(token, s, p);
            }
        }.token;
    }

    public IToken realpath(String path, final DoneRealPath done) {
        return new FileSystemCommand("realpath", new Object[]{ path }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                String p = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 2;
                    s = toFSError(args[0]);
                    if (s == null) p = (String)args[1];
                }
                done.doneRealPath(token, s, p);
            }
        }.token;
    }

    public IToken remove(String file_name, final DoneRemove done) {
        return new FileSystemCommand("remove", new Object[]{ file_name }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 1;
                    s = toFSError(args[0]);
                }
                done.doneRemove(token, s);
            }
        }.token;
    }

    public IToken rename(String old_path, String new_path, final DoneRename done) {
        return new FileSystemCommand("rename", new Object[]{ old_path, new_path }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 1;
                    s = toFSError(args[0]);
                }
                done.doneRename(token, s);
            }
        }.token;
    }

    public IToken rmdir(String path, final DoneRemove done) {
        return new FileSystemCommand("rmdir", new Object[]{ path }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 1;
                    s = toFSError(args[0]);
                }
                done.doneRemove(token, s);
            }
        }.token;
    }

    public IToken symlink(String link_path, String target_path, final DoneSymLink done) {
        return new FileSystemCommand("symlink", new Object[]{ link_path, target_path }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 1;
                    s = toFSError(args[0]);
                }
                done.doneSymLink(token, s);
            }
        }.token;
    }

    public IToken write(IFileHandle handle, long offset, byte[] data,
            int data_pos, int data_size, final DoneWrite done) {
        assert handle.getService() == this;
        String id = ((FileHandle)handle).id;
        return new FileSystemCommand("write", new Object[]{
                id, Long.valueOf(offset), new JSON.Binary(data, data_pos, data_size) }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 1;
                    s = toFSError(args[0]);
                }
                done.doneWrite(token, s);
            }
        }.token;
    }

    public IToken copy(String src_path, String dst_path,
            boolean copy_permissions, boolean copy_uidgid, final DoneCopy done) {
        return new FileSystemCommand("copy", new Object[]{
                src_path, dst_path, Boolean.valueOf(copy_permissions),
                Boolean.valueOf(copy_uidgid) }) {
            public void done(Exception error, Object[] args) {
                Status s = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 1;
                    s = toFSError(args[0]);
                }
                done.doneCopy(token, s);
            }
        }.token;
    }

    public IToken user(final DoneUser done) {
        return new FileSystemCommand("user", null) {
            @Override
            public void done(Exception error, Object[] args) {
                Status s = null;
                int r_uid = 0;
                int e_uid = 0;
                int r_gid = 0;
                int e_gid = 0;
                String home = null;
                if (error != null) {
                    s = new Status(error);
                }
                else {
                    assert args.length == 5;
                    r_uid = ((Number)args[0]).intValue();
                    e_uid = ((Number)args[1]).intValue();
                    r_gid = ((Number)args[2]).intValue();
                    e_gid = ((Number)args[3]).intValue();
                    home = (String)args[4];
                }
                done.doneUser(token, s, r_uid, e_uid, r_gid, e_gid, home);
            }
        }.token;
    }

    public String getName() {
        return NAME;
    }
    
    private Object toObject(FileAttrs attrs) {
        if (attrs == null) return null;
        Map<String,Object> m = new HashMap<String,Object>();
        if (attrs.attributes != null) m.putAll(attrs.attributes);
        if ((attrs.flags & ATTR_SIZE) != 0) {
            m.put("Size", Long.valueOf(attrs.size));
        }
        if ((attrs.flags & ATTR_UIDGID) != 0) {
            m.put("UID", Integer.valueOf(attrs.uid));
            m.put("GID", Integer.valueOf(attrs.gid));
        }
        if ((attrs.flags & ATTR_PERMISSIONS) != 0) {
            m.put("Permissions", Integer.valueOf(attrs.permissions));
        }
        if ((attrs.flags & ATTR_ACMODTIME) != 0) {
            m.put("ATime", Long.valueOf(attrs.atime));
            m.put("MTime", Long.valueOf(attrs.mtime));
        }
        return m;
    }
    
    @SuppressWarnings("unchecked")
    private FileAttrs toFileAttrs(Object o) {
        if (o == null) return null;
        Map<String,Object> m = new HashMap<String,Object>((Map<String,Object>)o);
        int flags = 0;
        long size = 0;
        int uid = 0;
        int gid = 0;
        int permissions = 0;
        long atime = 0;
        long mtime = 0;
        Number n = (Number)m.remove("Size");
        if (n != null) {
            size = n.longValue();
            flags |= ATTR_SIZE;
        }
        Number n1 = (Number)m.remove("UID");
        Number n2 = (Number)m.remove("GID");
        if (n1 != null && n2 != null) {
            uid = n1.intValue();
            gid = n2.intValue();
            flags |= ATTR_UIDGID;
        }
        n = (Number)m.remove("Permissions");
        if (n != null) {
            permissions = n.intValue();
            flags |= ATTR_PERMISSIONS;
        }
        n1 = (Number)m.remove("ATime");
        n2 = (Number)m.remove("MTime");
        if (n1 != null && n2 != null) {
            atime = n1.longValue();
            mtime = n2.longValue();
            flags |= ATTR_ACMODTIME;
        }
        return new FileAttrs(flags, size, uid, gid, permissions, atime, mtime, m);
    }
    
    private FileHandle toFileHandle(Object o) {
        if (o == null) return null;
        return new FileHandle(o.toString());
    }
    
    @SuppressWarnings("unchecked")
    private DirEntry[] toDirEntryArray(Object o) {
        if (o == null) return null;
        Collection<Map<String,Object>> c = (Collection<Map<String,Object>>)o;
        DirEntry[] res = new DirEntry[c.size()];
        int i = 0;
        for (Map<String,Object> m : c) {
            res[i++] = new DirEntry(
                    (String)m.get("FileName"),
                    (String)m.get("LongName"),
                    toFileAttrs(m.get("Attrs")));
        }
        return res;
    }
}

File: python/src/tcf/services/remote/FileSystemProxy.py
# *******************************************************************************
# * Copyright (c) 2011 Wind River Systems, Inc. and others.
# * All rights reserved. This program and the accompanying materials
# * are made available under the terms of the Eclipse Public License v1.0
# * which accompanies this distribution, and is available at
# * http://www.eclipse.org/legal/epl-v10.html
# *
# * Contributors:
# *     Wind River Systems - initial API and implementation
# *******************************************************************************

from tcf import errors, channel
from tcf.services import filesystem
from tcf.channel.Command import Command

class Status(filesystem.FileSystemException):
    def __init__(self, status_or_exception, message=None, attrs=None):
        if isinstance(status_or_exception, int):
            super(Status, self).__init__(message)
            self.status = status_or_exception
            self.attrs = attrs
        else:
            super(Status, self).__init__(status_or_exception)
            self.status = errors.TCF_ERROR_OTHER
            self.attrs = {}

    def getStatus(self):
        return self.status

    def getErrorCode(self):
        return self.attrs.get(errors.ERROR_CODE, 0)

    def getAltCode(self):
        return self.attrs.get(errors.ERROR_ALT_CODE, 0)

    def getAltOrg(self):
        return self.attrs.get(errors.ERROR_ALT_ORG)

    def getAttributes(self):
        return self.attrs

class FileSystemCommand(Command):
    def __init__(self, service, command, args):
        super(FileSystemCommand, self).__init__(service.channel, service, command, args)

    def _toSFError(self, data):
        if data is None: return None
        error_code = map.get(errors.ERROR_CODE)
        cmd = self.getCommandString()
        if len(cmd) > 72: cmd = cmd[0, 72] + "..."
        s = Status(error_code,
                "TCF command exception:" +
                "\nCommand: " + cmd +
                "\nException: " + self.toErrorString(data) +
                "\nError code: " + error_code, map)
        caused_by = map.get(errors.ERROR_CAUSED_BY)
        if caused_by is not None: s.initCause(self.toError(caused_by, False))
        return s

class FileSystemProxy(filesystem.FileSystemService):
    def __init__(self, channel):
        self.channel = channel

    def close(self, handle, done):
        assert handle.getService() is self
        id = handle.id
        service = self
        class CloseCommand(FileSystemCommand):
            def __init__(self):
                super(CloseCommand, self).__init__(service, "close", (id,))
            def done(self, error, args):
                s = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 1
                    s = self._toSFError(args[0])
                done.doneClose(self.token, s)
        return CloseCommand().token

    def setstat(self, path, attrs, done):
        dt = _toObject(attrs)
        service = self
        class SetStatCommand(FileSystemCommand):
            def __init__(self):
                super(SetStatCommand, self).__init__(service, "setstat", (path, dt))
            def done(self, error, args):
                s = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 1
                    s = self._toSFError(args[0])
                done.doneSetStat(self.token, s)
        return SetStatCommand().token

    def fsetstat(self, handle, attrs, done):
        assert handle.getService() is self
        id = handle.id
        dt = _toObject(attrs)
        service = self
        class FSetStatCommand(FileSystemCommand):
            def __init__(self):
                super(FSetStatCommand, self).__init__(service, "fsetstat", (id, dt))
            def done(self, error, args):
                s = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 1
                    s = self._toSFError(args[0])
                done.doneSetStat(self.token, s)
        return FSetStatCommand().token

    def stat(self, path, done):
        service = self
        class StatCommand(FileSystemCommand):
            def __init__(self):
                super(StatCommand, self).__init__(service, "stat", (path,))
            def done(self, error, args):
                s = None
                a = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 2
                    s = self._toSFError(args[0])
                    if not s: a = _toFileAttrs(args[1])
                done.doneStat(self.token, s, a)
        return StatCommand().token
    
    def fstat(self, handle, done):
        assert handle.getService() is self
        id = handle.id
        service = self
        class FStatCommand(FileSystemCommand):
            def __init__(self):
                super(FStatCommand, self).__init__(service, "fstat", (id,))
            def done(self, error, args):
                s = None
                a = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 2
                    s = self._toSFError(args[0])
                    if not s: a = _toFileAttrs(args[1])
                done.doneStat(self.token, s, a)
        return FStatCommand().token

    def lstat(self, path, done):
        service = self
        class LStatCommand(FileSystemCommand):
            def __init__(self):
                super(LStatCommand, self).__init__(service, "lstat", (path,))
            def done(self, error, args):
                s = None
                a = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 2
                    s = self._toSFError(args[0])
                    if not s: a = _toFileAttrs(args[1])
                done.doneStat(self.token, s, a)
        return LStatCommand().token

    def mkdir(self, path, attrs, done):
        dt = _toObject(attrs)
        service = self
        class MkDirCommand(FileSystemCommand):
            def __init__(self):
                super(MkDirCommand, self).__init__(service, "mkdir", (dt,))
            def done(self, error, args):
                s = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 1
                    s = self._toSFError(args[0])
                done.doneMkDir(self.token, s)
        return MkDirCommand().token

    def open(self, file_name, flags, attrs, done):
        dt = _toObject(attrs)
        service = self
        class OpenCommand(FileSystemCommand):
            def __init__(self):
                super(OpenCommand, self).__init__(service, "open", (file_name, flags, dt))
            def done(self, error, args):
                s = None
                h = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 2
                    s = self._toSFError(args[0])
                    if not s: h = self._toFileHandle(args[1])
                done.doneOpen(self.token, s, h)
        return OpenCommand().token

    def opendir(self, path, done):
        service = self
        class OpenDirCommand(FileSystemCommand):
            def __init__(self):
                super(OpenDirCommand, self).__init__(service, "opendir", (path,))
            def done(self, error, args):
                s = None
                h = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 2
                    s = self._toSFError(args[0])
                    if not s: h = self._toFileHandle(args[1])
                done.doneOpen(self.token, s, h)
        return OpenDirCommand().token

    def read(self, handle, offset, len, done):
        assert handle.getService() is self
        id = handle.id
        service = self
        class ReadCommand(FileSystemCommand):
            def __init__(self):
                super(ReadCommand, self).__init__(service, "read", (id, offset, len))
            def done(self, error, args):
                s = None
                b = None
                eof = False
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 3
                    s = self._toSFError(args[1])
                    if not s:
                        b = channel.toByteArray(args[0])
                        eof = args[2]
                done.doneRead(self.token, s, b, eof)
        return ReadCommand().token

    def readdir(self, handle, done):
        assert handle.getService() is self
        id = handle.id
        service = self
        class ReadDirCommand(FileSystemCommand):
            def __init__(self):
                super(ReadDirCommand, self).__init__(service, "readdir", (id,))
            def done(self, error, args):
                s = None
                b = None
                eof = False
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 3
                    s = self._toSFError(args[1])
                    if not s:
                        b = _toDirEntryArray(args[0])
                        eof = args[2]
                done.doneReadDir(self.token, s, b, eof)
        return ReadDirCommand().token

    def roots(self, done):
        service = self
        class RootCommand(FileSystemCommand):
            def __init__(self):
                super(RootCommand, self).__init__(service, "roots", None)
            def done(self, error, args):
                s = None
                b = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 2
                    s = self._toSFError(args[1])
                    if not s:
                        b = _toDirEntryArray(args[0])
                done.doneRoots(self.token, s, b)
        return RootCommand().token

    def readlink(self, path, done):
        service = self
        class ReadLinkCommand(FileSystemCommand):
            def __init__(self):
                super(ReadLinkCommand, self).__init__(service, "readlink", (path,))
            def done(self, error, args):
                s = None
                p = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 2
                    s = self._toSFError(args[0])
                    if not s:
                        p = args[1]
                done.doneReadLink(self.token, s, p)
        return ReadLinkCommand().token

    def realpath(self, path, done):
        service = self
        class RealPathCommand(FileSystemCommand):
            def __init__(self):
                super(RealPathCommand, self).__init__(service, "realpath", (path,))
            def done(self, error, args):
                s = None
                p = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 2
                    s = self._toSFError(args[0])
                    if not s:
                        p = args[1]
                done.doneRealPath(self.token, s, p)
        return RealPathCommand().token

    def remove(self, file_name, done):
        service = self
        class RemoveCommand(FileSystemCommand):
            def __init__(self):
                super(RemoveCommand, self).__init__(service, "remove", (file_name,))
            def done(self, error, args):
                s = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 1
                    s = self._toSFError(args[0])
                done.doneRemove(self.token, s)
        return RemoveCommand().token

    def rename(self, old_path, new_path, done):
        service = self
        class RenameCommand(FileSystemCommand):
            def __init__(self):
                super(RenameCommand, self).__init__(service, "rename", (old_path, new_path))
            def done(self, error, args):
                s = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 1
                    s = self._toSFError(args[0])
                done.doneRename(self.token, s)
        return RenameCommand().token

    def rmdir(self, path, done):
        service = self
        class RmDirCommand(FileSystemCommand):
            def __init__(self):
                super(RmDirCommand, self).__init__(service, "rmdir", (path,))
            def done(self, error, args):
                s = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 1
                    s = self._toSFError(args[0])
                done.doneRemove(self.token, s)
        return RmDirCommand().token

    def symlink(self, link_path, target_path, done):
        service = self
        class SymLinkCommand(FileSystemCommand):
            def __init__(self):
                super(SymLinkCommand, self).__init__(service, "symlink", (link_path, target_path))
            def done(self, error, args):
                s = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 1
                    s = self._toSFError(args[0])
                done.doneSymLink(self.token, s)
        return SymLinkCommand().token

    def write(self, handle, offset, data, data_pos, data_size, done):
        assert handle.getService() is self
        id = handle.id
        binary = bytearray(data[data_pos:data_pos+data_size])
        service = self
        class WriteCommand(FileSystemCommand):
            def __init__(self):
                super(WriteCommand, self).__init__(service, "write", (id, offset, binary))
            def done(self, error, args):
                s = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 1
                    s = self._toSFError(args[0])
                done.doneWrite(self.token, s)
        return WriteCommand().token

    def copy(self, src_path, dst_path, copy_permissions, copy_uidgid, done):
        service = self
        class CopyCommand(FileSystemCommand):
            def __init__(self):
                super(CopyCommand, self).__init__(service, "copy", 
                        (id, src_path, dst_path, copy_permissions, copy_uidgid))
            def done(self, error, args):
                s = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 1
                    s = self._toSFError(args[0])
                done.doneCopy(self.token, s)
        return CopyCommand().token

    def user(self, done):
        service = self
        class UserCommand(FileSystemCommand):
            def __init__(self):
                super(UserCommand, self).__init__(service, "user", None)
            def done(self, error, args):
                s = None
                r_uid = 0
                e_uid = 0
                r_gid = 0
                e_gid = 0
                home = None
                if error:
                    s = Status(error)
                else:
                    assert len(args) == 5
                    r_uid, e_uid, r_gid, e_gid, home = args
                done.doneUser(self.token, s, r_uid, e_uid, r_gid, e_gid, home)
        return UserCommand().token

    def _toFileHandle(self, o):
        if o is None: return None
        return filesystem.FileHandle(self, o)

def _toObject(attrs):
    if attrs is None: return None
    m = {}
    if attrs.attributes is not None: m.update(attrs.attributes)
    if (attrs.flags & filesystem.ATTR_SIZE) != 0:
        m.put("Size", attrs.size)
    if (attrs.flags & filesystem.ATTR_UIDGID) != 0:
        m.put("UID", attrs.uid)
        m.put("GID", attrs.gid)
    if (attrs.flags & filesystem.ATTR_PERMISSIONS) != 0:
        m.put("Permissions", attrs.permissions)
    if (attrs.flags & filesystem.ATTR_ACMODTIME) != 0:
        m.put("ATime", attrs.atime)
        m.put("MTime", attrs.mtime)
    return m

def _toFileAttrs(m):
    if m is None: return None
    flags = 0
    size = 0
    uid = 0
    gid = 0
    permissions = 0
    atime = 0
    mtime = 0
    n = m.pop("Size", None)
    if n is not None:
        size = n
        flags |= filesystem.ATTR_SIZE
    n1 = m.pop("UID", None)
    n2 = m.pop("GID", None)
    if n1 is not None and n2 is not None:
        uid = n1
        gid = n2
        flags |= filesystem.ATTR_UIDGID
    n = m.pop("Permissions", None)
    if n is not None:
        permissions = n
        flags |= filesystem.ATTR_PERMISSIONS
    n1 = m.pop("ATime", None)
    n2 = m.pop("MTime", None)
    if n1 is not None and n2 is not None:
        atime = n1
        mtime = n2
        flags |= filesystem.ATTR_ACMODTIME
    return filesystem.FileAttrs(flags, size, uid, gid, permissions, atime, mtime, m)

def _toDirEntryArray(o):
    if o is None: return None
    res = []
    for m in o:
        entry = filesystem.DirEntry(m.get("FileName"), m.get("LongName"), _toFileAttrs(m.get("Attrs")))
        res.append(entry)
    return res
